### ISSUE TITLE
Questions about training a model

### ISSUE THREAD

**Issue by CatarinaPC** [explosion/spaCy#4886]
I had a couple questions I hope someone can answer (also hope I'm posting this at the right place or with the right tag):

1. Should one use cli or [python code](https://spacy.io/usage/training#example-train-ner) to train a model (more specifically a NER model)? Using python code seems more complicated. Are there any advantages to it?

2. I read that hyper-parameters can be changed when using cli, is it possible to do the same but in python? Or is that one of the disadvantages regarding the first question?

3. Is there any way to perform hyper-parameter tuning/selection? I couldn't find anything regarding this. if not, is it possible to use another library combined with SpaCy to achieve this?

4. What is the purpose of the dev data used with the train command on the cli? Or, what are the changes that are made at each iteration in which the dev data is used to evaluate?

## Your Environment

* **spaCy version:** 2.2.3
* **Platform:** Linux-5.0.0-37-generic-x86_64-with-Ubuntu-18.04-bionic
* **Python version:** 3.6.8

**Comment by adrianeboyd** [comment]
The python code from the docs is mainly just a demo example that shows the basics of a training loop and the train CLI is better for training models for real use. It's used internally to train all the models spacy distributes. Not all parameters are available as CLI options, so you may want to add additional options or modify the script as you need:

https://github.com/explosion/spaCy/blob/master/spacy/cli/train.py

Spacy doesn't have any built-in tools for hyperparameter tuning and it's also currently not as easy as it should be to modify the model parameters. (There are major improvements in progress for the next version of thinc.)

In the train CLI the dev data is used to choose the best model out of all the iterations and optionally for early stopping.

**Comment by CatarinaPC** [comment]
Thank you so much for your answer.

I hope you can clarify some things for me:

> Not all parameters are available as CLI options

So there are more parameters that can be changed besides the ones in [this](https://spacy.io/api/cli#train-hyperparams) list?

> the models spacy distributes

> not as easy as it should be to modify the model parameters

Do the models SpaCy distributes use the default hyperparameters?

I am trying to train my own NER model. From your last message I think the best route is to do it via CLI. Do you think it is worth or significant trying to change the hyperparameters? 

And do you think it is worth or even possible to put some sort of "tuning"/"selection", using other libraries or cross-validation, over the output of the CLI?

**Comment by adrianeboyd** [comment]
> I hope you can clarify some things for me:
> 
> > Not all parameters are available as CLI options
> 
> So there are more parameters that can be changed besides the ones in [this](https://spacy.io/api/cli#train-hyperparams) list?

Sorry, I was thinking of CLI arguments from the section above that one. Setting environment variables is currently the best way to set those hyperparameters.

> Do the models SpaCy distributes use the default hyperparameters?

Yes. I think the only non-default value for the v2.2 models is `-ovl 0.3`.

> I am trying to train my own NER model. From your last message I think the best route is to do it via CLI. Do you think it is worth or significant trying to change the hyperparameters?
> 
> And do you think it is worth or even possible to put some sort of "tuning"/"selection", using other libraries or cross-validation, over the output of the CLI?

It's definitely possible to use another library in combination with the train CLI. By default, each model directory saved by the CLI includes a `meta.json` with the model performance saved under the key "accuracy". You could also easily modify the script to return or save the results in an alternate format that's easier to work with. I haven't experimented much in this area, though, so I don't have a good feeling for how much hyperparameter tuning might help overall.
